README
Group menmber: Derrick Wright  dwright@iastate.edu
               Hamza Shahid    hmshahid@iastate.edu
               Sonyta Ung      ung@iastate.edu
The process to achieve this Parsing Monster Definitions are:

- making a monster object class and making important attributes to save info to the class
- make dice class with attributes for upcoming assignments
- parsing through the txt file printing attributes monsters have and their respective discriptions
- saving information we parse through to use for later




change_log
Change Log-

2/08/2021 5pm made function to make the 21x80 grid 
2/08/2021 5pm wrote method to generate random number of rooms with random y and x values (within a range) 
2/9/2021 6pm - fixed bugs that caused the rooms to be too thin or too wide -- improved randomization 
2/10/2021 - fixed bugs that caused rooms to print over each other and clip outside of map boundaries
2/10/2021 5pm -made staircases 1 up and 1 down random based on room placement
2/11/2021 5-8 pm wrote method for calculating and placing down corridors
2/12/2021 5pm: decided to use professor’s code drop
2/12/2021 5pm: started writing write method for the save switch inputs for file size and file type markers.
2/14/2021 4 pm: began writing writefile method for dungeon map
2/16/2021 5-8pm continued to writefile method for dungeon map
2/17/2021 5pm: started writing a read file method for loading the dungeon map.
2/18/2021 5 pm: finished writefile method for dungeon map and fixed various conversion errors
2/22/2021 5pm: created non tunneling monster map struct and made a function incorporating professors algorithm to find shortest path to character 
2/22/2021 5pm: created tunneling monster map structure and incorporated professors code to calculate distance between all nodes
2/23/2021 5pm: created comparable for non tunneling monster and printing out costs to get to monster from each room, stair, corridor 
2/24/2021 6pm: realized that distances were wrong for tunneling monster, utilized monster1 code to fix discrepancies
3/2/2021 5pm made intelligent monster and sub types of that
3/3/2021 5pm finished intelligent monster and sub types 
3/4/2021 started priority queue 
3/10/2021: worked on making user interface for pc to move (ncurses)
3/11/2021: completed user interface to get pc to move
3/11/2021: completed user interface to print infomationion
3/24/2021: added keys to toggle between fog view and normal
3/24/2021: added teleport functionality 
3/25/2021: added targeting mode, added goto and r  keys to toggle, testing
03/28/2021: discuss / brainstorming how to handle the assignment
03/29/2021: create monster maker class, dice class and implement them
03/30/2021: implement parser to read monster description
03/31/2021: implement monster_color and testing
04/01/2021: implement monster ability and testing

move.cpp
#include "move.h"

#include <unistd.h>
#include <stdlib.h>
#include <assert.h>

#include "dungeon.h"
#include "heap.h"
#include "move.h"
#include "npc.h"
#include "pc.h"

#include "character.h"
#include "utils.h"
#include "path.h"
#include "event.h"
#include "io.h"
#include "ncurses.h"


int y;
int j;
bool q = true;
void do_combat(dungeon_t *d, character_t *atk, character_t *def)
{
  int can_see_atk, can_see_def;
  const char *organs[] = {
    "liver",                   /*  0 */
    "pancreas",                /*  1 */
    "heart",                   /*  2 */
    "eye",                     /*  3 */
    "arm",                     /*  4 */
    "leg",                     /*  5 */
    "intestines",              /*  6 */
    "gall bladder",            /*  7 */
    "lungs",                   /*  8 */
    "hand",                    /*  9 */
    "foot",                    /* 10 */
    "spinal cord",             /* 11 */
    "pituitary gland",         /* 12 */
    "thyroid",                 /* 13 */
    "tongue",                  /* 14 */
    "bladder",                 /* 15 */
    "diaphram",                /* 16 */
    "stomach",                 /* 17 */
    "pharynx",                 /* 18 */
    "esophagus",               /* 19 */
    "trachea",                 /* 20 */
    "urethra",                 /* 21 */
    "spleen",                  /* 22 */
    "ganglia",                 /* 23 */
    "ear",                     /* 24 */
    "subcutaneous tissue"      /* 25 */
    "cerebellum",              /* 26 */ /* Brain parts begin here */
    "hippocampus",             /* 27 */
    "frontal lobe",            /* 28 */
    "brain",                   /* 29 */
  };
  int part;

  if (def->alive) {
    def->alive = 0;
    charpair(def->position) = NULL;
    
    if (def != &d->pc) {
      d->num_monsters--;
    } else {
      if ((part = rand() % (sizeof (organs) / sizeof (organs[0]))) < 26) {
        io_queue_message("As the %c eats your %s, "
                         "you wonder if there is an afterlife.",
                         atk->symbol, organs[part]);
      } else {
        io_queue_message("Your last thoughts fade away as "
                         "the %c eats your %s...",
                         atk->symbol, organs[part]);
      }
      /* Queue an empty message, otherwise the game will not pause for *
       * player to see above.                                          */
      io_queue_message("");
    }
    atk->kills[kill_direct]++;
    atk->kills[kill_avenged] += (def->kills[kill_direct] +
                                  def->kills[kill_avenged]);
  }

  if (atk == &d->pc) {
    io_queue_message("You smite the %c!", def->symbol);
  }

  can_see_atk = can_see(d, &d->pc, atk);
  can_see_def = can_see(d, &d->pc, def);

  if (atk != &d->pc && def != &d->pc) {
    if (can_see_atk && !can_see_def) {
      io_queue_message("The %c callously murders some poor, "
                       "defenseless creature.", atk->symbol);
    }
    if (can_see_def && !can_see_atk) {
      io_queue_message("Something kills the helpless %c.", def->symbol);
    }
    if (can_see_atk && can_see_def) {
      io_queue_message("You watch in abject horror as the %c "
                       "gruesomely murders the %c!", atk->symbol, def->symbol);
    }
  }
}

void move_character(dungeon_t *d, character_t *c, pair_t next)
{
  if (charpair(next) &&
      ((next[dim_y] != c->position[dim_y]) ||
       (next[dim_x] != c->position[dim_x]))) {
    do_combat(d, c, charpair(next));
  } else {
    /* No character in new position. */

    d->character[c->position[dim_y]][c->position[dim_x]] = NULL;
    c->position[dim_y] = next[dim_y];
    c->position[dim_x] = next[dim_x];
    d->character[c->position[dim_y]][c->position[dim_x]] = c;
  }
}
int check(int f){
  j = f;
  return j;
}
int check2(int v){
  y = v;
  return y;
}
void do_moves(dungeon_t *d)
{
  
  
    
  pair_t next;
  character_t *c;
  event_t *e;

  /* Remove the PC when it is PC turn.  Replace on next call.  This allows *
   * use to completely uninit the heap when generating a new level without *
   * worrying about deleting the PC.                                       */

  if (pc_is_alive(d) ) {
    /* The PC always goes first one a tie, so we don't use new_event().  *
     * We generate one manually so that we can set the PC sequence       *
     * number to zero.                                                   */
    e = static_cast<event_t*>(malloc(sizeof (*e)));
    e->type = event_character_turn;
    /* Hack: New dungeons are marked.  Unmark and ensure PC goes at d->time, *
     * otherwise, monsters get a turn before the PC.                         */
    if (d->is_new) {
      d->is_new = 0;
      e->time = d->time;
    } else {
      e->time = d->time + (1000 / d->pc.speed);
    }
    e->sequence = 0;
    e->c = &d->pc;
    heap_insert(&d->events, e);
  }
  if(check(j)%2!=0){
     q = false;

  }
  if(check(j)%2==0){
     q = true;

  }
  if(q){

  while (pc_is_alive(d) &&
         (e = (event_t*)heap_remove_min(&d->events)) &&
         ((e->type != event_character_turn) || (e->c !=&d->pc)) ) {
    d->time = e->time;
    if (e->type == event_character_turn) {
      c = e->c;
    }
    if (!c->alive) {
      if (d->character[c->position[dim_y]][c->position[dim_x]] == c) {
        d->character[c->position[dim_y]][c->position[dim_x]] = NULL;
      }
      if (c != &d->pc) {
        event_delete(e);
      }
      continue;
    }
    npc_next_pos(d, c, next);
    move_character(d, c, next);
     heap_insert(&d->events, update_event(d, e, 1000 / c->speed));
    
    }
    

    
  }
   
  
 if(check2(y) %2 != 0){
   io_display_teleport(d);
 }
 if(check2(y)%2==0){
  io_display_with_fog(d);
 }
  
  if (pc_is_alive(d) && e->c == &d->pc) {
    c = e->c;
    d->time = e->time;
    /* Kind of kludgey, but because the PC is never in the queue when   *
     * we are outside of this function, the PC event has to get deleted *
     * and recreated every time we leave and re-enter this function.    */
    e->c = NULL;
    
    event_delete(e);
    
   
    
     
    io_handle_input(d);
    
  }
}

void dir_nearest_wall(dungeon_t *d, character_t *c, pair_t dir)
{
  dir[dim_x] = dir[dim_y] = 0;

  if (c->position[dim_x] != 1 && c->position[dim_x] != DUNGEON_X - 2) {
    dir[dim_x] = (c->position[dim_x] > DUNGEON_X - c->position[dim_x] ? 1 : -1);
  }
  if (c->position[dim_y] != 1 && c->position[dim_y] != DUNGEON_Y - 2) {
    dir[dim_y] = (c->position[dim_y] > DUNGEON_Y - c->position[dim_y] ? 1 : -1);
  }
}

uint32_t against_wall(dungeon_t *d, character_t *c)
{
  return ((mapxy(c->position[dim_x] - 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x] + 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] - 1) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] + 1) == ter_wall_immutable));
}

uint32_t in_corner(dungeon_t *d, character_t *c)
{
  uint32_t num_immutable;

  num_immutable = 0;

  num_immutable += (mapxy(c->position[dim_x] - 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x] + 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] - 1) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] + 1) == ter_wall_immutable);

  return num_immutable > 1;
}

static void new_dungeon_level(dungeon_t *d, uint32_t dir)
{
  /* Eventually up and down will be independantly meaningful. *
   * For now, simply generate a new dungeon.                  */

  switch (dir) {
  case '<':
  case '>':
    new_dungeon(d);
    break;
  default:
    break;
  }
}


uint32_t move_pc(dungeon_t *d, uint32_t dir)
{
  pair_t next;
  uint32_t was_stairs = 0;
  const char *wallmsg[] = {
    "There's a wall in the way.",
    "BUMP!",
    "Ouch!",
    "You stub your toe.",
    "You can't go that way.",
    "You admire the engravings.",
    "Are you drunk?"
  };

  next[dim_y] = d->pc.position[dim_y];
  next[dim_x] = d->pc.position[dim_x];


  switch (dir) {
  case 1:
  case 2:
  case 3:
    next[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    next[dim_y]--;
    break;
  }
  switch (dir) {
  case 1:
  case 4:
  case 7:
    next[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    next[dim_x]++;
    break;
  case '<':
    if (mappair(d->pc.position) == ter_stairs_up) {
      was_stairs = 1;
      new_dungeon_level(d, '<');
    }
    break;
  case '>':
    if (mappair(d->pc.position) == ter_stairs_down) {
      was_stairs = 1;
      new_dungeon_level(d, '>');
    }
    break;
  }

  if (was_stairs) {
    return 0;
  }

  if ((dir != '>') && (dir != '<') && (mappair(next) >= ter_floor)) {
    move_character(d, &d->pc, next);
    dijkstra(d);
    dijkstra_tunnel(d);

    return 0;
  } else if (mappair(next) < ter_floor) {
    io_queue_message(wallmsg[rand() % (sizeof (wallmsg) /
                                       sizeof (wallmsg[0]))]);
   
  }
  
  

  return 1;
}
